### ISSUE TITLE
grpc (C++) server close connection for stream window overflow

### ISSUE THREAD

**Issue by yancl** [grpc/grpc-go#1192]
Hi,
  I have a grpc **c++** server and a grpc **go** client, they communicate with each other by a **bi-direction stream**. when the go client send some(about 1k) messages(4k payload each), the client got some error:
`2017/04/19 12:36:15 transport: http2Client.notifyError got notified that the client transport was broken EOF.`

and through tcpdump i find that the server close the connection(port 50051 is the server):
```
12:36:15.165077 IP 127.0.0.1.56209 > 127.0.0.1.50051: Flags [P.], seq 1810591:1815067, ack 2687, win 434, options [nop,nop,TS val 7919327 ecr 7919326], length 4476
12:36:15.166217 IP 127.0.0.1.50051 > 127.0.0.1.56209: Flags [F.], seq 2687, ack 1815067, win 19309, options [nop,nop,TS val 7919327 ecr 7919327], length 0
12:36:15.166249 IP 127.0.0.1.56209 > 127.0.0.1.50051: Flags [.], seq 1815067:1880550, ack 2688, win 434, options [nop,nop,TS val 7919327 ecr 7919327], length 65483
12:36:15.166649 IP 127.0.0.1.56209 > 127.0.0.1.50051: Flags [P.], seq 1880550:1916099, ack 2688, win 434, options [nop,nop,TS val 7919328 ecr 7919327], length 35549
12:36:15.166898 IP 127.0.0.1.50051 > 127.0.0.1.56209: Flags [R], seq 1317965835, win 0, length 0
```

then i found some debug log from grpc server:
```
D0419 12:36:15.165398502   22422 chttp2_transport.c:2154]    set connectivity_state=4
D0419 12:36:15.165401804   22422 connectivity_state.c:184]   SET: 0x7fe200001438 server_transport: READY --> SHUTDOWN [close_transport] error=0x7fe20c01f780 {"created":"@1492605375.128074698","description":"Delayed close due to in-progress write","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.c","file_line":521,"referenced_errors":[{"created":"@1492605375.128072799","description":"Failed parsing HTTP/2","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.c","file_line":1971,"grpc_status":14,"referenced_errors":[{"created":"@1492605375.128069267","description":"frame of size 4467 overflows incoming window of 911","file":"src/core/ext/transport/chttp2/transport/parsing.c","file_line":412}]}]}
D0419 12:36:15.166190700   22422 connectivity_state.c:211]   NOTIFY: 0x7fe200001438 server_transport: 0x7fe200005350
```

according the log, the server says that:`frame of size 4467 overflows incoming window of 911`,
so is this a  bug or with mis-configed params?

NOTE: at first i think it was a grpc server bug, but when i make same requests with grpc c++ client, there isn't any problem. and when using the grpc golang client, it trigger the problem very easy.

so will you please fix it?

**Comment by yancl** [comment]
How to produce:

```
// start server
cd grpc/examples/cpp/route_guide
GRPC_TRACE=all GRPC_VERBOSITY=DEBUG ./route_guide_server


// edit client
cd google.golang.org/grpc/examples/route_guide
edit client/client.go
 89 // runRecordRoute sends a sequence of points to server and expects to get a RouteSummary from server.
 90 func runRecordRoute(client pb.RouteGuideClient) {
 91     // Create a random number of random points
 92     r := rand.New(rand.NewSource(time.Now().UnixNano()))
 93     //pointCount := int(r.Int31n(100)) + 2 // Traverse at least two points
 94     pointCount = 10 * 1024 * 1024 //emulate a lot of message

// run client
go run client/client.go
```

then you will find the following errors in client:
```
2017/04/20 08:12:01 Traversing 10485760 points.
2017/04/20 08:12:02 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/04/20 08:12:02 &{0xc4200746e0}.Send(latitude:520000000 longitude:500000000 ) = rpc error: code = Internal desc = transport is closing
exit status 1
```

and you will find the following errors in server:
```
D0420 08:12:02.682009547   14262 chttp2_transport.c:2154]    set connectivity_state=4
D0420 08:12:02.682035245   14262 connectivity_state.c:184]   SET: 0x7f7d440014c8 server_transport: READY --> SHUTDOWN [close_transport] error=0x7f7d442ca150 {"created":"@1492675922.681983109","description":"Delayed close due to in-progress write","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.c","file_line":521,"referenced_errors":[{"created":"@1492675922.681977493","description":"Failed parsing HTTP/2","file":"src/core/ext/transport/chttp2/transport/chttp2_transport.c","file_line":1971,"grpc_status":14,"referenced_errors":[{"created":"@1492675922.681963731","description":"frame of size 27 overflows incoming window of 3","file":"src/core/ext/transport/chttp2/transport/parsing.c","file_line":412}]}]}
D0420 08:12:02.682038018   14262 connectivity_state.c:211]   NOTIFY: 0x7f7d440014c8 server_transport: 0x7f7d44005360
```

when make the same change with the grpc c++ client, no problem appears.

@iamqizhao @menghanl  any thoughts? thanks.

**Comment by yancl** [comment]
and i add a go client with go server test.  no problem appears
```
google.golang.org/grpc/examples/route_guide$ go run server/server.go

google.golang.org/grpc/examples/route_guide$ go run client/client.go
2017/04/20 08:51:03 Traversing 10485760 points.
2017/04/20 08:57:16 Route summary: point_count:10485760 distance:1483317118 elapsed_time:372
```

**so it seems like the problem occurs between go and c++.**

**Comment by yancl** [comment]
the c++ client with golang server, no problem appears
```
grpc/examples/cpp/route_guide$ ./route_guide_client
DB parsed, loaded 100 features.
Finished trip with 10485760 points
Passed 10485760 features
Travelled 733387647 meters
It took 355 seconds
```

**Comment by mehrdada** [comment]
I am unable to reproduce this problem on `master` for gRPC C++ & Go. Seems to work fine on my machine (`Ubuntu 16.04.2 LTS, Linux 4.8.0-45-generic #48~16.04.1-Ubuntu SMP Fri Mar 24 12:46:56 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux, Go 1.8.1, gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609`). Which versions of gRPC C++ and Go are you using and on what platform?

```
~/go/src/google.golang.org/grpc/examples/route_guide$ go run client/client.go --server_addr localhost:50051
2017/04/26 20:16:36 Getting feature for point (409146138, -746188906)
2017/04/26 20:16:36 location:<latitude:409146138 longitude:-746188906 >
2017/04/26 20:16:36 Getting feature for point (0, 0)
2017/04/26 20:16:36 location:<>
2017/04/26 20:16:36 Looking for features within lo:<latitude:400000000 longitude:-750000000 > hi:<latitude:420000000 longitude:-730000000 >
2017/04/26 20:16:41 Traversing 10485760 points.
2017/04/26 20:18:20 Route summary: point_count:10485760 distance:-1619001344 elapsed_time:98
2017/04/26 20:18:20 Got message First message at point(0, 1)
2017/04/26 20:18:20 Got message Second message at point(0, 2)
2017/04/26 20:18:20 Got message Third message at point(0, 3)
```

**Comment by yancl** [comment]
@mehrdada, thank you for your reply. 
When i start a new clean ubuntu vm(`Linux ip-10-1-7-106 3.13.0-48-generic #80-Ubuntu SMP Thu Mar 12 11:16:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`) and use the master branch of gRPC c++&golang code. The problem does not occur easily without tracing open. so i think maybe the c++ server consumes messages too fast. then i add a `sleep(10)` to the c++ server handler to slow down the message consumption, and the problem occurs easily now.

```
~/grpc/examples/cpp/route_guide$ ./route_guide_server

//code changed
133   Status RecordRoute(ServerContext* context, ServerReader<Point>* reader,
134                      RouteSummary* summary) override {
135     Point point;
136     int point_count = 0;
137     int feature_count = 0;
138     float distance = 0.0;
139     Point previous;
140
141     system_clock::time_point start_time = system_clock::now();
142     while (reader->Read(&point)) {
             // to slow down the process
143       printf("sleep 10 seconds\n");
144       sleep(10);
145       point_count++;
146       if (!GetFeatureName(point, feature_list_).empty()) {
147         feature_count++;
148       }
149       if (point_count != 1) {
150         distance += GetDistance(previous, point);
151       }
152       previous = point;
153     }
```

```
/go/src/google.golang.org/grpc/examples/route_guide/client$ go run client.go

//errors from golang client
2017/04/29 05:44:58 Traversing 10485760 points.
2017/04/29 05:45:28 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/04/29 05:45:28 &{0xc42006e580}.Send(latitude:-680000000 longitude:-830000000 ) = rpc error: code = Internal desc = transport is closing
exit status 1
```

**Comment by mehrdada** [comment]
I'm still investigating the issue to get at the root cause. It is a bit more nuanced than initially expected.

**Comment by yancl** [comment]
Thank you very very much for that @mehrdada .
Currently we downgrade grpc(c++) from **1.2.0** to **1.0.0** to workaround the problem and it works well as i know.

**Comment by mehrdada** [comment]
@yancl Have you downgraded both the server and the client or are you just downgrading the C++ server?

**Comment by mehrdada** [comment]
This issue has been super hard to debug and needs more work, but just to give you an update, at this point it sounds like the problem might be in the `http2` package's handling of settings frames appropriately.

**Comment by yancl** [comment]
@mehrdada i downgraded the grpc c++ server code while using the latest grpc golang client code. well, it is just as a note and may be not that useful.

Yes, thank you for your update and hard work:)

**Comment by mehrdada** [comment]
It seems #1262 has the same root cause and https://github.com/grpc/grpc/commits/80ccebb2e16355666809eb5368751683149eb6a8 on the C repo is the commit beyond which the bug is manifested clearly. However, simply undoing the commit on master is not helping, which indicates the root cause is probably somewhere else and that commit just makes it more easily visible.

**Comment by mehrdada** [comment]
Integer overflow bug fixed by #1262. Please let us know if you see similar issues.

**Comment by yancl** [comment]
Thank you very much @mehrdada :)
